vllm/v1/core/specialized_manager.py

```diff
@@ -34,14 +46,173 @@ def __init__(
         # Needs special handling for find_longest_cache_hit if eagle is enabled
         self.use_eagle = use_eagle
 
+        # Mapping from request ID to blocks to track the blocks allocated
+        # for each request, so that we can free the blocks when the request
+        # is finished.
+        self.req_to_blocks: defaultdict[str,
+                                        list[KVCacheBlock]] = defaultdict(list)
+
+        # {req_id: The number of cached blocks for this given request}
+        # This is used to track the number of cached blocks for each request.
+        # This is only used to track the RUNNING requests, we do not track the
+        # data for reempted ones.
+        self.num_cached_block: dict[str, int] = {}
+        self.max_num_blocks_per_req = cdiv(max_model_len, self.block_size)
+
+        self.num_kv_cache_groups = num_kv_cache_groups
+        self.caching_hash_fn = caching_hash_fn
+
+    def get_num_blocks_to_allocate(
+            self, request_id: str, num_tokens: int,
+            new_computed_blocks: list[KVCacheBlock]) -> int:
+        """
+        Get the number of blocks needed to be allocated for the request.
+
+        Args:
+            request_id: The request ID.
+            num_tokens: The total number of tokens that need a slot (including 
+                tokens that are already allocated).
+            new_computed_blocks: The new computed blocks just hitting the
+                prefix caching.
+
+        Returns:
+            The number of blocks.
+        """
+
+        num_required_blocks = cdiv(num_tokens, self.block_size)
+        num_new_blocks = num_required_blocks - len(new_computed_blocks) - \
+            len(self.req_to_blocks[request_id])
```

Comment:
style nit: Can we not use `\`?
```suggestion
        num_new_blocks = (num_required_blocks - len(new_computed_blocks) -
            len(self.req_to_blocks[request_id]))
```